extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/DatasourceOperations.java

```diff
@@ -173,23 +190,82 @@ public int executeUpdate(String query) throws SQLException {
    * @throws SQLException : Exception caught during transaction execution.
    */
   public void runWithinTransaction(TransactionCallback callback) throws SQLException {
-    try (Connection connection = borrowConnection()) {
-      boolean autoCommit = connection.getAutoCommit();
-      connection.setAutoCommit(false);
-      boolean success = false;
+    withRetries(
+        () -> {
+          try (Connection connection = borrowConnection()) {
+            boolean autoCommit = connection.getAutoCommit();
+            boolean success = false;
+            connection.setAutoCommit(false);
+            try {
+              try (Statement statement = connection.createStatement()) {
+                success = callback.execute(statement);
+              }
+            } finally {
+              if (success) {
+                connection.commit();
+              } else {
+                connection.rollback();
+              }
+              connection.setAutoCommit(autoCommit);
+            }
+          }
+          return null;
+        });
+  }
+
+  private boolean isRetryable(SQLException e) {
+    String sqlState = e.getSQLState();
+
+    if (sqlState != null) {
+      return sqlState.equals(DEADLOCK_SQL_CODE)
+          || // Deadlock detected
+          sqlState.equals(SERIALIZATION_FAILURE_SQL_CODE); // Serialization failure
+    }
+
+    // Additionally, one might check for specific error messages or other conditions
+    return e.getMessage().contains("connection refused")
+        || e.getMessage().contains("connection reset");
+  }
+
+  public <T> T withRetries(Operation<T> operation) throws SQLException {
+    int attempts = 0;
+    // maximum number of retries.
+    int maxAttempts = relationalJdbcConfiguration.maxRetries().orElse(1);
```

Comment:
IMHO, max attempts should not be restricted by default (MAX_VALUE). Only time should be restricted.

Reply:
Here is my take on this, Since we are randomizing the retry duration, we strictly can't restrict number of retries ? if i am constrained by lets say number of concurrent connection having no control on retries count might leave this a bit unpredictable for planning the concurrent connection we wanna support, hence bounded by maxRetries 

This serves as a middle ground, helps both the user who wanna rely just on max retries (by setting max duration to something very big) or retry just on max duration or have combination of both.

Reply:
I doubt it's a good "middle ground". Time is most important here, which is not being controlled here - it can run for hours and the client's already hung up.

Reply:
I see, Here was my thought process, that those people who have configured just the maxRetries be bounded by default max retry duration ? 5000 ms in that sense before we sleep for ages we do check before our sleep that if that sleep is well within our maxRetry duration ? if the person configures infinite retry lets says INT_MAX and the duration INT_MAX then its their responsibility ?